File: src/test/java/org/trp/shincolle/entity/base/path/ShipLegacyNavigationPolicyTest.java
package org.trp.shincolle.entity.base.path;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

class ShipLegacyNavigationPolicyTest {

    @Test
    void sameNavigationTargetShouldUseLegacyDriftTolerance() {
        ShipLegacyNavigationPolicy.Target origin = new ShipLegacyNavigationPolicy.Target(0, 64, 0);

        assertTrue(ShipLegacyNavigationPolicy.isSameNavigationTarget(
                origin, new ShipLegacyNavigationPolicy.Target(3, 64, 0)));
        assertFalse(ShipLegacyNavigationPolicy.isSameNavigationTarget(
                origin, new ShipLegacyNavigationPolicy.Target(4, 64, 0)));
        assertFalse(ShipLegacyNavigationPolicy.isSameNavigationTarget(null, origin));
    }

    @Test
    void stuckProgressShouldResetOnlyForFreshPathsOrRealTargetChanges() {
        assertTrue(ShipLegacyNavigationPolicy.shouldResetStuckProgress(false, true));
        assertTrue(ShipLegacyNavigationPolicy.shouldResetStuckProgress(true, false));
        assertFalse(ShipLegacyNavigationPolicy.shouldResetStuckProgress(true, true));
    }

    @Test
    void setPathLogsShouldTriggerForFailureChangeTargetChangeOrThrottleExpiry() {
        ShipLegacyNavigationPolicy.Target target = new ShipLegacyNavigationPolicy.Target(8, 64, 8);
        ShipLegacyNavigationPolicy.Target nearby = new ShipLegacyNavigationPolicy.Target(10, 64, 8);
        ShipLegacyNavigationPolicy.Target far = new ShipLegacyNavigationPolicy.Target(12, 64, 8);

        assertTrue(ShipLegacyNavigationPolicy.shouldLogSetPath(false, true, target, target, 0, 0));
        assertTrue(ShipLegacyNavigationPolicy.shouldLogSetPath(false, false, target, far, 0, 0));
        assertFalse(ShipLegacyNavigationPolicy.shouldLogSetPath(false, false, target, nearby,
                ShipLegacyNavigationPolicy.NAVIGATION_SET_PATH_LOG_INTERVAL - 1, 0));
        assertTrue(ShipLegacyNavigationPolicy.shouldLogSetPath(false, false, target, nearby,
                ShipLegacyNavigationPolicy.NAVIGATION_SET_PATH_LOG_INTERVAL, 0));
    }

    @Test
    void navigationEventLogsShouldRespectTargetDriftAndThrottle() {
        ShipLegacyNavigationPolicy.Target target = new ShipLegacyNavigationPolicy.Target(8, 64, 8);
        ShipLegacyNavigationPolicy.Target nearby = new ShipLegacyNavigationPolicy.Target(10, 64, 8);
        ShipLegacyNavigationPolicy.Target far = new ShipLegacyNavigationPolicy.Target(12, 64, 8);

        assertTrue(ShipLegacyNavigationPolicy.shouldLogNavigationEvent(target, target, 0, Integer.MIN_VALUE));
        assertTrue(ShipLegacyNavigationPolicy.shouldLogNavigationEvent(target, far, 0, 0));
        assertFalse(ShipLegacyNavigationPolicy.shouldLogNavigationEvent(target, nearby,
                ShipLegacyNavigationPolicy.NAVIGATION_DEBUG_LOG_INTERVAL - 1, 0));
        assertTrue(ShipLegacyNavigationPolicy.shouldLogNavigationEvent(target, nearby,
                ShipLegacyNavigationPolicy.NAVIGATION_DEBUG_LOG_INTERVAL, 0));
    }

    @Test
    void timeoutPolicyShouldClampSpeedAndRetryAfterDoubleLimit() {
        assertEquals(6000.0D, ShipLegacyNavigationPolicy.calculateTimeoutLimit(1.0D, 0.0D));
        assertFalse(ShipLegacyNavigationPolicy.shouldRetryTimedOutPath(200L, 100.0D));
        assertTrue(ShipLegacyNavigationPolicy.shouldRetryTimedOutPath(201L, 100.0D));
        assertFalse(ShipLegacyNavigationPolicy.shouldRetryTimedOutPath(1L, 0.0D));
    }
}

File: src/test/kotlin/org/trp/shincolle/entity/base/path/ShipLegacyNavigationPolicyTest.kt
package org.trp.shincolle.entity.base.path

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class ShipLegacyNavigationPolicyTest {

    @Test
    fun `same navigation target should use legacy drift tolerance`() {
        val origin = ShipLegacyNavigationPolicy.Target(0, 64, 0)

        assertThat(
            ShipLegacyNavigationPolicy.isSameNavigationTarget(
                origin,
                ShipLegacyNavigationPolicy.Target(3, 64, 0)
            )
        ).isTrue()
        assertThat(
            ShipLegacyNavigationPolicy.isSameNavigationTarget(
                origin,
                ShipLegacyNavigationPolicy.Target(4, 64, 0)
            )
        ).isFalse()
        assertThat(ShipLegacyNavigationPolicy.isSameNavigationTarget(null, origin)).isFalse()
    }

    @Test
    fun `stuck progress should reset only for fresh paths or real target changes`() {
        assertThat(ShipLegacyNavigationPolicy.shouldResetStuckProgress(false, true)).isTrue()
        assertThat(ShipLegacyNavigationPolicy.shouldResetStuckProgress(true, false)).isTrue()
        assertThat(ShipLegacyNavigationPolicy.shouldResetStuckProgress(true, true)).isFalse()
    }

    @Test
    fun `set path logs should trigger for failure change target change or throttle expiry`() {
        val target = ShipLegacyNavigationPolicy.Target(8, 64, 8)
        val nearby = ShipLegacyNavigationPolicy.Target(10, 64, 8)
        val far = ShipLegacyNavigationPolicy.Target(12, 64, 8)

        assertThat(ShipLegacyNavigationPolicy.shouldLogSetPath(false, true, target, target, 0, 0)).isTrue()
        assertThat(ShipLegacyNavigationPolicy.shouldLogSetPath(false, false, target, far, 0, 0)).isTrue()
        assertThat(
            ShipLegacyNavigationPolicy.shouldLogSetPath(
                false,
                false,
                target,
                nearby,
                ShipLegacyNavigationPolicy.NAVIGATION_SET_PATH_LOG_INTERVAL - 1,
                0
            )
        ).isFalse()
        assertThat(
            ShipLegacyNavigationPolicy.shouldLogSetPath(
                false,
                false,
                target,
                nearby,
                ShipLegacyNavigationPolicy.NAVIGATION_SET_PATH_LOG_INTERVAL,
                0
            )
        ).isTrue()
    }

    @Test
    fun `navigation event logs should respect target drift and throttle`() {
        val target = ShipLegacyNavigationPolicy.Target(8, 64, 8)
        val nearby = ShipLegacyNavigationPolicy.Target(10, 64, 8)
        val far = ShipLegacyNavigationPolicy.Target(12, 64, 8)

        assertThat(ShipLegacyNavigationPolicy.shouldLogNavigationEvent(target, target, 0, Int.MIN_VALUE)).isTrue()
        assertThat(ShipLegacyNavigationPolicy.shouldLogNavigationEvent(target, far, 0, 0)).isTrue()
        assertThat(
            ShipLegacyNavigationPolicy.shouldLogNavigationEvent(
                target,
                nearby,
                ShipLegacyNavigationPolicy.NAVIGATION_DEBUG_LOG_INTERVAL - 1,
                0
            )
        ).isFalse()
        assertThat(
            ShipLegacyNavigationPolicy.shouldLogNavigationEvent(
                target,
                nearby,
                ShipLegacyNavigationPolicy.NAVIGATION_DEBUG_LOG_INTERVAL,
                0
            )
        ).isTrue()
    }

    @Test
    fun `timeout policy should clamp speed and retry after double limit`() {
        assertThat(ShipLegacyNavigationPolicy.calculateTimeoutLimit(1.0, 0.0)).isEqualTo(6000.0)
        assertThat(ShipLegacyNavigationPolicy.shouldRetryTimedOutPath(200L, 100.0)).isFalse()
        assertThat(ShipLegacyNavigationPolicy.shouldRetryTimedOutPath(201L, 100.0)).isTrue()
        assertThat(ShipLegacyNavigationPolicy.shouldRetryTimedOutPath(1L, 0.0)).isFalse()
    }
}
